### ISSUE TITLE
Creating an Event Definition breaks entire Alerts & Events section

### ISSUE THREAD

**Issue by em-ex** [Graylog2/graylog2-server#20098]
## Expected Behavior
Creating an Event Definition creates a single item in the list which works and can be edited or removed

## Current Behavior
Creating an Event Definition breaks Alerts & Events in a way that we cannot see any items in the list anymore as it hangs up and then throws a `FetchError` with message saying:
> There was an error fetching a resource: Gateway Time-out

Both "Alerts & Events" and "Event Definitions" tabs stop working, only "Notifications" continue to work.
Graylog log fills with the following messages:
```log
ERROR: org.graylog.scheduler.JobSchedulerService - Error running job execution engine
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Null jobDefinitionType
Caused by: java.lang.NullPointerException: Null jobDefinitionType
```
<details>
<summary>Entire error log</summary>

```log
2024-08-06 16:48:06,531 ERROR: org.graylog.scheduler.JobSchedulerService - Error running job execution engine
java.lang.RuntimeException: IOException encountered while parsing
    at org.mongojack.internal.stream.JacksonDecoder.decode(JacksonDecoder.java:45) ~[graylog.jar:?]
    at org.mongojack.internal.stream.JacksonCodec.decode(JacksonCodec.java:75) ~[graylog.jar:?]
		
at com.mongodb.internal.operation.CommandResultDocumentCodec.readValue(CommandResultDocumentCodec.java:58) ~[graylog.jar:?]
    at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:87) ~[graylog.jar:?]
    at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:42) ~[graylog.jar:?]
    at com.mongodb.internal.connection.ReplyMessage.<init>(ReplyMessage.java:45) ~[graylog.jar:?]
    at com.mongodb.internal.connection.InternalStreamConnection.getCommandResult(InternalStreamConnection.java:544) ~[graylog.jar:?]
    at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:438) ~[graylog.jar:?]
    at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:354) ~[graylog.jar:?]
    at com.mongodb.internal.connection.UsageTrackingInternalConnection.sendAndReceive(UsageTrackingInternalConnection.java:114) ~[graylog.jar:?]
    at com.mongodb.internal.connection.DefaultConnectionPool$PooledConnection.sendAndReceive(DefaultConnectionPool.java:743) ~[graylog.jar:?]
    at com.mongodb.internal.connection.CommandProtocolImpl.execute(CommandProtocolImpl.java:76) ~[graylog.jar:?]
    at com.mongodb.internal.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:209) ~[graylog.jar:?]
    at com.mongodb.internal.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:115) ~[graylog.jar:?]
    at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:83) ~[graylog.jar:?]
    at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:74) ~[graylog.jar:?]
    at com.mongodb.internal.connection.DefaultServer$OperationCountTrackingConnection.command(DefaultServer.java:299) ~[graylog.jar:?]
    at com.mongodb.internal.operation.SyncOperationHelper.lambda$executeRetryableWrite$9(SyncOperationHelper.java:240) ~[graylog.jar:?]
    at com.mongodb.internal.operation.SyncOperationHelper.lambda$withSourceAndConnection$0(SyncOperationHelper.java:124) ~[graylog.jar:?]
    at com.mongodb.internal.operation.SyncOperationHelper.withSuppliedResource(SyncOperationHelper.java:149) ~[graylog.jar:?]
    at com.mongodb.internal.operation.SyncOperationHelper.lambda$withSourceAndConnection$1(SyncOperationHelper.java:123) ~[graylog.jar:?]
    at com.mongodb.internal.operation.SyncOperationHelper.withSuppliedResource(SyncOperationHelper.java:149) ~[graylog.jar:?]
    at com.mongodb.internal.operation.SyncOperationHelper.withSourceAndConnection(SyncOperationHelper.java:122) ~[graylog.jar:?]
    at com.mongodb.internal.operation.SyncOperationHelper.lambda$executeRetryableWrite$10(SyncOperationHelper.java:226) ~[graylog.jar:?]
    at com.mongodb.internal.operation.SyncOperationHelper.lambda$decorateWriteWithRetries$11(SyncOperationHelper.java:280) ~[graylog.jar:?]
    at com.mongodb.internal.async.function.RetryingSyncSupplier.get(RetryingSyncSupplier.java:67) ~[graylog.jar:?]
    at com.mongodb.internal.operation.SyncOperationHelper.executeRetryableWrite(SyncOperationHelper.java:252) ~[graylog.jar:?]
    at com.mongodb.internal.operation.BaseFindAndModifyOperation.execute(BaseFindAndModifyOperation.java:80) ~[graylog.jar:?]
    at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:173) ~[graylog.jar:?]
    at com.mongodb.client.internal.MongoCollectionImpl.executeFindOneAndUpdate(MongoCollectionImpl.java:760) ~[graylog.jar:?]
    at com.mongodb.client.internal.MongoCollectionImpl.findOneAndUpdate(MongoCollectionImpl.java:740) ~[graylog.jar:?]
    at org.mongojack.MongoCollectionDecorator.findOneAndUpdate(MongoCollectionDecorator.java:931) ~[graylog.jar:?]
    at org.mongojack.JacksonDBCollection.findAndModify(JacksonDBCollection.java:293) ~[graylog.jar:?]
    at org.graylog.scheduler.DBJobTriggerService.nextRunnableTrigger(DBJobTriggerService.java:380) ~[graylog.jar:?]
    at org.graylog.scheduler.JobExecutionEngine.execute(JobExecutionEngine.java:198) ~[graylog.jar:?]
    at org.graylog.scheduler.JobSchedulerService.run(JobSchedulerService.java:120) ~[graylog.jar:?]
    at com.google.common.util.concurrent.AbstractExecutionThreadService$1.lambda$doStart$1(AbstractExecutionThreadService.java:57) ~[graylog.jar:?]
    at com.google.common.util.concurrent.Callables.lambda$threadRenaming$3(Callables.java:105) ~[graylog.jar:?]
		
at java.base/java.lang.Thread.run(Unknown Source) [?:?]
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Null jobDefinitionType
 at [Source: (String); byte offset: #UNKNOWN] (through reference chain: org.graylog.scheduler.AutoValue_JobTriggerDto$Builder["job_definition_type"])
    at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:276) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:630) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:618) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeSetAndReturn(MethodProperty.java:173) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.deser.BuilderBasedDeserializer.vanillaDeserialize(BuilderBasedDeserializer.java:294) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.deser.BuilderBasedDeserializer.deserialize(BuilderBasedDeserializer.java:218) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:342) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.ObjectReader._bind(ObjectReader.java:2099) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1249) ~[graylog.jar:?]
    at org.mongojack.internal.stream.JacksonDecoder.decode(JacksonDecoder.java:43) ~[graylog.jar:?]
    ... 38 more
Caused by: java.lang.NullPointerException: Null jobDefinitionType
    at org.graylog.scheduler.AutoValue_JobTriggerDto$Builder.jobDefinitionType(AutoValue_JobTriggerDto.java:337) ~[graylog.jar:?]
    at jdk.internal.reflect.GeneratedMethodAccessor183.invoke(Unknown Source) ~[?:?]
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:?]
    at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[?:?]
    at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeSetAndReturn(MethodProperty.java:170) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.deser.BuilderBasedDeserializer.vanillaDeserialize(BuilderBasedDeserializer.java:294) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.deser.BuilderBasedDeserializer.deserialize(BuilderBasedDeserializer.java:218) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:342) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.ObjectReader._bind(ObjectReader.java:2099) ~[graylog.jar:?]
    at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1249) ~[graylog.jar:?]
    at org.mongojack.internal.stream.JacksonDecoder.decode(JacksonDecoder.java:43) ~[graylog.jar:?]
    ... 38 more
```
</details>

## Steps to Reproduce (for bugs)
1. Create a new Event Definition
2. Witness "Alerts & Events" fail to work afterwards 

## Context
We just want to send out email notifications on certain critical errors. What are we missing?

## Your Environment
* Graylog Version: 6.0.4 (still, the same behavior observed on 4.3.15 and then 5.0.13, prior to 6.x upgrade)
* Java Version: Eclipse Adoptium 17.0.11
* OpenSearch Version: still ElasticSearch 7.10.2
* MongoDB Version: 6.0.16
* Operating System: Ubuntu 20.04.6 
* Browser version: Firefox 128

**Comment by janheise** [comment]
@em-ex Hi, can you be more specific about the steps you're doing?

You observed the same behaviour on 4.3.15 and 5.0.13 - so both times you started with a working installation, both tabs in the Frontend show existing Event Definitions? And then, when you add another one, it starts to fail? Or do you start from scratch, with no definitions at all?

I'm creating those objects quite often but have never experienced your problem. Creation of an Event Definition is a multi-step process. What do you exactly do? Can you make some screenshots of the definition you try to create?

Do you use any 3rd party plugins?

**Comment by em-ex** [comment]
Thank you @janheise for taking time to look into this!

Forgive me if I omitted any valuable information – I took over this project only recently and am trying to make it make sense myself. So far, I've only managed to upgrade our GL stack and now trying to figure out this critical error email notification assignment.

We do start from scratch – no Event Definitions exist yet. Once we add an event definition (of Filter & Aggregation type, a simple query to find certain log message during a single day) and save it, the list of existing definitions won't render at all, resulting in "Gateway Time-out" error message popping on the bottom of the screen. I'm only able to restore the functionality by removing the offending event definition under mongo shell.

`db.getCollection("scheduler_triggers").distinct('job_definition_type')` returns empty array `[]` – that's not a good sign, right?

**Comment by janheise** [comment]
@em-ex Hi again, the empty collection does not sound right. 

So, what are you removing in Mongo if you remove the offending definition? The trigger in `scheduler_triggers` and the event definition in `scheduler_job_definitions`?

per default there should be at least one Event definition in the system <img width="1028" alt="Screenshot 2024-08-06 at 20 56 58" src="https://github.com/user-attachments/assets/0c855b41-0701-4ac0-9d57-513d08061294">
If it's missing, you broke your installation. imho.

You could check out one of our docker examples at e.g. https://github.com/Graylog2/docker-compose/blob/main/enterprise/docker-compose.yml and just check that an Event Definition creates just fine. If it still fails, it's a bug ;-) - if not, your setup seems to be broken.

**Comment by em-ex** [comment]
No, I'm removing my event definition from `db.event_definitions` at index 1. Index 0 is occupied by "System notification events", which you've just mentioned above, and it's visible in web GUI as well. 

Just to clarify: you want me to deploy a factory-defaults GL instance in docker on the same machine to test if it works as is, right?

<details><summary>PS. My event definition in mongo looks like this</summary>

```js
  {
    _id: ObjectId('66b250990999100b65461ef4'),
    _scope: 'DEFAULT',
    title: 'Application error',
    description: '',
    updated_at: ISODate('2024-08-06T16:34:33.851Z'),
    priority: 2,
    alert: true,
    config: {
      type: 'aggregation-v1',
      query: 'message:"application error"',
      query_parameters: [],
      filters: [],
      streams: [],
      group_by: [],
      series: [],
      conditions: { expression: null },
      search_within_ms: Long('86400000'),
      execute_every_ms: Long('300000'),
      event_limit: 1
    },
    field_spec: {},
    key_spec: [],
    notification_settings: { grace_period_ms: Long('3600000'), backlog_size: Long('0') },
    notifications: [
      {
        notification_id: '66915d3be682a659489df113',
        notification_parameters: null
      }
    ],
    storage: [
      {
        type: 'persist-to-streams-v1',
        streams: [ '000000000000000000000002' ]
      }
    ],
    state: 'ENABLED'  
  }
```

Slightly redacted, but in essence it's the same thing as in production.
</details>

**Comment by janheise** [comment]
@em-ex I'd like to check if you can reproduce the problem with a vanilla setup - or if it's really specific to your configuration. It's not even necessary that you install it on the machine that you run your setup. Can be your local machine, too (so you don't mess with your prod at all).

From the stacktrace: your event definition in the corresponding mongo collection seems not to be the problem but the scheduler entries/job triggers. 
You ran a query above on the triggers that returned an empty array. Is the collection empty?
Also, how many objects in `scheduler_job_definitions` do you have that contain `{"type": "event-processor-execution-v1"`in the config field?

**Comment by em-ex** [comment]
I went to check what's in the `scheduler_job_definitions` collection and I found 4 items from my previous attempts to make the event definition to work. I removed them, and now there are two: 
- one of type `notification-execution-v1` that corresponds with our custom email notification
- one of type `check-for-cert-renewal-execution-v1` (from default GL config?)

Removing my orphaned scheduler job definitions made the "Alerts & Events" and "Event Definitions" tabs work again. Once I'd recreated my event definition and run it, the tabs ceased to work properly again. Close, but no cigar.

I then went to check the `scheduler_triggers` collection, and it's pretty packed, despite none of them having a distinct `job_definition_type` (if the query I ran yesterday is to be believed):
```
graylog> db.scheduler_triggers.count()
2331983
```
Given that you haven't experienced my problem by performing essentially the same steps, and that there are no similar issues on GitHub, I suspect that the problem is related to our configuration after all. Is there a point in running an empty GL instance?

The main mystery to me that I can't quite wrap my head around is why are we getting the gateway timeout message – where does it even come from?

**Comment by janheise** [comment]
@em-ex the gateway timeout message comes up because the frontend never receives any data from the backend to show (because the backend fails deserialising the objects) and is not relevant in this context.

The error is that there probably already are objects inside the `scheduler_triggers` collection that create the error - or worse creating new events create new erroneous objects in the `scheduler_triggers` collection (but I have no idea how the latter can happen, given that the other things you already checked look fine imho).
The former is not easy to fix because of the high number of objects.

**Comment by em-ex** [comment]
Since we had no event definitions... uhm... defined by our team, I went and nuked the entire `scheduler_triggers` collection. I made a backup of the entire VM and stopped the GL container beforehand, of course. 
I then restarted GL and created the same event definition and it just seems to work ¯\\_(ツ)_/¯
![image](https://github.com/user-attachments/assets/1bbe12dc-c808-48cd-9694-4023a506dbd3)
Before this radical decision, I snooped around in mongo db and the oldest entry in `scheduler_triggers` collection had `job_definition_type: null`. And I could've started by removing only those, but felt ~~cute~~ courageous, so... the question, could there be anything important there? (I can still salvage it from the backup db)
